File: algorithm/Programmers/JAVA/LV3/taxi_fare.java
package Programmers.JAVA.LV3;

import java.util.*;

public class taxi_fare {
    static List<node>[] list;
    static boolean check[];
    static int[] distance;
    static int[] a_distance;
    static int[] b_distance;

    public static void main(String[] args) {
        int n = 6; // 지점갯수
        int s = 4; // 출발지
        int a = 6; // a도착지
        int b = 2; // b도착지
        int[][] fares = { { 4, 1, 10 }, { 3, 5, 24 }, { 5, 6, 2 }, { 3, 1, 41 }, { 5, 1, 24 }, { 4, 6, 50 },
                { 2, 4, 66 }, { 2, 3, 22 }, { 1, 6, 25 } };

        solution(n, s, a, b, fares);

    }

    private static int solution(int n, int s, int a, int b, int[][] fares) {
        int answer = Integer.MAX_VALUE;

        list = new ArrayList[n + 1];
        check = new boolean[n + 1];
        distance = new int[n + 1];
        a_distance = new int[n + 1];
        b_distance = new int[n + 1];

        Arrays.fill(distance, Integer.MAX_VALUE);
        Arrays.fill(a_distance, Integer.MAX_VALUE);
        Arrays.fill(b_distance, Integer.MAX_VALUE);

        for (int i = 0; i < n + 1; i++) {
            list[i] = new ArrayList<>();
        }

        for (int i = 0; i < fares.length; i++) {
            list[fares[i][0]].add(new node(fares[i][1], fares[i][2]));
            list[fares[i][1]].add(new node(fares[i][0], fares[i][2]));
        }

        distance = djikstra(s, distance);
        a_distance = djikstra(a, a_distance);
        b_distance = djikstra(b, b_distance);

        for (int i = 1; i <= n; i++)
            answer = Math.min(answer, distance[i] + a_distance[i] + b_distance[i]);

        return answer;
    }

    private static int[] djikstra(int start, int[] distance) {
        PriorityQueue<node> q = new PriorityQueue<>();

        Arrays.fill(distance, Integer.MAX_VALUE);

        q.add(new node(start, 0));
        distance[start] = 0;

        while (!q.isEmpty()) {
            node n = q.poll();
            int now = n.index;

            if (check[now] == true)
                continue;
            check[now] = true;

            for (node t : list[now]) {
                if (distance[t.index] > distance[now] + t.weight) {
                    distance[t.index] = distance[now] + t.weight;
                    q.add(new node(t.index, distance[t.index]));
                }
            }
        }

        Arrays.fill(check, false);
        return distance;
    }
}

class node implements Comparable<node> {
    int index;
    int weight;

    node(int i, int w) {
        this.index = i;
        this.weight = w;
    }

    @Override
    public int compareTo(node o) {
        return weight - o.weight;
    }
}

File: algorithm/Programmers/JAVA/LV3/Kotlin/taxi_fare.kt
import java.util.*
import kotlin.collections.ArrayList

lateinit var list: Array<ArrayList<node>>
lateinit var check: Array<Boolean>
lateinit var distance: Array<Int>
lateinit var a_distance: Array<Int>
lateinit var b_distance: Array<Int>

fun solution(n: Int, s: Int, a: Int, b: Int, fares: Array<IntArray>): Int {
    var answer: Int = Integer.MAX_VALUE

    list = Array(n + 1) { ArrayList<node>() }
    check = Array(n + 1) { false }
    distance = Array(n + 1) { Int.MAX_VALUE }
    a_distance = Array(n + 1) { Int.MAX_VALUE }
    b_distance = Array(n + 1) { Int.MAX_VALUE }

    for (i in 0 until n + 1) {
        list[i] = ArrayList()
    }

    for (i in 0 until fares.size) {
        list[fares[i][0]].add(node(fares[i][1], fares[i][2]))
        list[fares[i][1]].add(node(fares[i][0], fares[i][2]))
    }

    distance = djikstra(s, distance)
    a_distance = djikstra(a, a_distance)
    b_distance = djikstra(b, b_distance)

    for (i in 1..n) {
        answer = Math.min(answer, distance[i] + a_distance[i] + b_distance[i])
    }

    return answer
}

fun djikstra(start: Int, distance: Array<Int>): Array<Int> {
    var q = PriorityQueue<node>()

    Arrays.fill(distance, Int.MAX_VALUE)

    q.add(node(start, 0))
    distance[start] = 0

    while (!q.isEmpty()) {
        var n: node = q.poll()
        var now = n.index

        if (check[now] == true) continue
        check[now] = true

        for (t in list[now]) {
            if (distance[t.index] > distance[now] + t.weight) {
                distance[t.index] = distance[now] + t.weight
                q.add(node(t.index, distance[t.index]))
            }
        }
    }
    Arrays.fill(check, false)

    return distance
}

class node : Comparable<node> {

    var index = 0
    var weight = 0

    constructor(i: Int, w: Int) {
        this.index = i
        this.weight = w
    }

    override fun compareTo(other: node): Int {
        return weight - other.weight
    }
}
